x-pack/plugins/security/server/lib/authentication/authenticator.js

```diff
@@ -234,19 +234,22 @@ class Authenticator {
   /**
    * Returns provider iterator where providers are sorted in the order of priority (based on the session ownership).
    * @param {Object} sessionValue Current session value.
+   * @param {Hapi.Request} request HapiJS request instance.
    * @returns {Iterator.<Object>}
    */
-  *_providerIterator(sessionValue) {
-    // If there is no session to predict which provider to use first, let's use the order
-    // providers are configured in. Otherwise return provider that owns session first, and only then the rest
-    // of providers.
-    if (!sessionValue) {
+  *_providerIterator(sessionValue, request) {
+    // If there is no way to predict which provider to use first, let's use the order providers are configured in.
+    // Otherwise return provider that either owns session or can handle 3rd-party login request first, and only then
+    // the rest of providers.
+    const shouldHandleSAMLResponse = this._isSAMLResponse(request) && this._providers.has('saml');
+    if (!sessionValue && !shouldHandleSAMLResponse) {
       yield* this._providers;
     } else {
-      yield [sessionValue.provider, this._providers.get(sessionValue.provider)];
+      const provider = shouldHandleSAMLResponse ? 'saml' : sessionValue.provider;
```

Comment:
__note:__ this particular line can be reversed to prefer provider from the session instead that would prevent SAML login if user has an active session already (with basic/token provider) and require user to explcitly log out first. This may remove potential confusion when users semi-silently re-authenticate themselves (may have been solved with `overwritten_session` we have in more recent versions of Kibana).

But current solution feels a bit more convenient for the admins that would log in to Kibana from Cloud console and would know what they are doing :man_shrugging: What do you think?